buildSrc/src/main/kotlin/codegen/math/GenerateMath.kt

```diff
@@ -0,0 +1,1027 @@
+					ClassName(packageName, "${bitsType.simpleName}Vector$componentCount").let { bitsVectorType ->
+						extensionFunction(type.nestedClass("Companion"), "fromBits") {
+							parameter("bits", bitsVectorType)
+							returns(type)
+							val args = componentNames.joinToString(",\n\t", "\n\t", "\n") {
+								val unsignedConversion = if (unsigned) ".to${bitsType.simpleName.drop(1)}()" else ""
+								"${baseType!!.simpleName}.fromBits(bits.$it$unsignedConversion)"
+							}
+							statement("return %T($args)", type)
+						}
+
+						extensionFunction(mutableType.nestedClass("Companion"), "fromBits") {
+							parameter("bits", bitsVectorType)
+							returns(type)
+							val args = componentNames.joinToString(",\n\t", "\n\t", "\n") {
+								val unsignedConversion = if (unsigned) ".to${bitsType.simpleName.drop(1)}()" else ""
+								"${baseType!!.simpleName}.fromBits(bits.$it$unsignedConversion)"
+							}
+							statement("return %T($args)", type)
+						}
+					}
+				}
+
+				when (baseType) {
+					FLOAT -> {
+						fromBitsFunction(INT, false)
+						fromBitsFunction(U_INT, true)
+					}
+					DOUBLE -> {
+						fromBitsFunction(LONG, false)
+						fromBitsFunction(U_LONG, true)
+					}
+				}
+
+				// frexp
+
+				// ldexp
+
+			}.writeTo(commonDir.get().asFile)
+		}
+	}
+
+	private fun geometricFunctions() {
+		vectorTypes.forEach { (type, baseType, componentCount) ->
+			if (baseType == BOOLEAN) return@forEach
+
+			val componentNames = allComponentNames.take(componentCount!!)
+			baseType ?: error("null base type invalid for vector types")
+
+			buildFile(packageName, "${type.simpleName}Geometric") {
+				indent("\t")
+
+				// Dot Product
+
+				extensionFunction(type, "dot") {
+					modifiers(INFIX)
+					parameter("other", type)
+					returns(baseType)
+					val result = componentNames.joinToString(" + ") { "$it * other.$it" }
+					statement("return \n\t($result)")
+				}
+
+				// Cross Product
+
+				if (componentCount == 3) {
+					extensionFunction(type, "cross") {
+						modifiers(INFIX)
+						parameter("other", type)
+						returns(type)
+						val args = componentNames
+							.drop(1)
+							.plus(componentNames.take(2))
+							.zipWithNext()
+							.joinToString(",\n\t", "\n\t", "\n") { (c1, c2) ->
+								"$c1 * other.$c2 - other.$c1 * $c2"
+							}
+						statement("return %T($args)", type)
+					}
+				}
+			}.writeTo(commonDir.get().asFile)
+		}
+	}
+
+	private fun vectorTypes() {
+		vectorTypes.forEach { (type, baseType, componentCount) ->
+			val mutableVectorType = ClassName(packageName, "Mutable${type.simpleName}")
+			val componentNames = allComponentNames.take(componentCount!!)
+			baseType ?: error("null base type invalid for vectors")
+
+			val (zero) = literal("0", baseType, null)
+
+			buildFile(packageName, type.simpleName) {
+				indent("\t")
+
+				import("kotlin.math", "sqrt")
+
+				buildClass(type) {
+					modifiers(SEALED)
+
+					companionObject()
```

Comment:
Let's introduce immutable constants for each vector type. One for zero, one and each component where possible. Should save on allocation.
- `ZERO = Vector(0)`
- `ONE = Vector(1)`
- `X = Vector(1, 0, 0)`
- `Y = Vector(0, 1, 0)`
- `Z = Vector(0, 0, 1)`
- `W = Vector(0, 0, 0, 1)`

Reply:
sure